src/tracing/node_trace_writer.cc

```diff
@@ -0,0 +1,182 @@
+#include "tracing/node_trace_writer.h"
+
+#include <string.h>
+#include <fcntl.h>
+
+#include "util.h"
+
+namespace node {
+namespace tracing {
+
+NodeTraceWriter::NodeTraceWriter(uv_loop_t* tracing_loop)
+    : tracing_loop_(tracing_loop) {
+  flush_signal_.data = this;
+  int err = uv_async_init(tracing_loop_, &flush_signal_, FlushSignalCb);
+  CHECK_EQ(err, 0);
+
+  exit_signal_.data = this;
+  err = uv_async_init(tracing_loop_, &exit_signal_, ExitSignalCb);
+  CHECK_EQ(err, 0);
+}
+
+void NodeTraceWriter::WriteSuffix() {
+  // If our final log file has traces, then end the file appropriately.
+  // This means that if no trace events are recorded, then no trace file is
+  // produced.
+  bool should_flush = false;
+  {
+    Mutex::ScopedLock scoped_lock(stream_mutex_);
+    if (total_traces_ > 0) {
+      total_traces_ = 0; // so we don't write it again in FlushPrivate
+      // Appends "]}" to stream_.
+      delete json_trace_writer_;
+      should_flush = true;
+    }
+  }
+  if (should_flush) {
+    Flush(true);
+  }
+}
+
+NodeTraceWriter::~NodeTraceWriter() {
+  WriteSuffix();
+  uv_fs_t req;
+  int err;
+  if (fd_ != -1) {
+    err = uv_fs_close(tracing_loop_, &req, fd_, nullptr);
+    CHECK_EQ(err, 0);
+    uv_fs_req_cleanup(&req);
+  }
+  uv_async_send(&exit_signal_);
+  Mutex::ScopedLock scoped_lock(request_mutex_);
+  while(!exited_) {
+    exit_cond_.Wait(scoped_lock);
```

Comment:
The best solution I can come up with is to add a `Close()` method to `NodeTraceWriter` and `NodeTraceBuffer`, so that the async handles can be closed without deleting those objects. Then they can be safely deleted after the loop thread exits. I'm working on a patch.